DROP SUBSCRIPTION query called on invalid retention policy corrupts database

It seems like I accidentally corrupted the database by running a `DROP subscription` query while playing around with kapacitor.

- CentOS release 6.6 (Final), 2.6.32-504.16.2.el6.x86_64
- InfluxDB version 0.9.6.1-1 installed from yum

We installed kapacitor and started it with this config.

    [influxdb]
      # Connect to an InfluxDB cluster
      # Kapacitor can subscribe, query and write to this cluster.
      # Using InfluxDB is not required and can be disabled.
      enabled = true
      urls = ["http://localhost:8086"]
      username = "telegraf"
      password = "<%= scope.lookupvar('influx::params::telegraf_password') %>"
      timeout = 0
      # Subscriptions use the UDP network protocl.
      # The following options of for the created UDP listeners for each subscription.
      # Number of packets to buffer when reading packets off the socket.
      udp-buffer = 1000
      # The size in bytes of the OS read buffer for the UDP socket.
      # A value of 0 indicates use the OS default.
      udp-read-buffer = 0

      [influxdb.subscriptions]
        # Set of databases and retention policies to subscribe to.
        # If empty will subscribe to all, minus the list in
        # influxdb.excluded-subscriptions
        #
        # Format
        # db_name = <list of retention policies>
        #
        # Example:
        # my_database = [ "default", "longterm" ]
      [influxdb.excluded-subscriptions]
        # Set of databases and retention policies to exclude from the subscriptions.
        # If influxdb.subscriptions is empty it will subscribe to all
        # except databases listed here.
        #
        # Format
        # db_name = <list of retention policies>
        #
        # Example:
        # my_database = [ "default", "longterm" ]

Since we didn't specify any subscriptions, kapacitor set up the following:

    curl -G http://localhost:8086/query --data-urlencode "q=SHOW subscriptions"

    name: _internal
    ---------------
    retention_policy    name        mode    destinations
    monitor         kapacitor   ANY [udp://localhost:35359]


    name: telegraf
    --------------
    retention_policy    name        mode    destinations
    default         kapacitor   ANY [udp://localhost:50576]

I changed the kapacitor config to only have the subscription `telegraf = ["default"]`, and figured I should delete the old subscription on `_internal`:

    curl -G http://localhost:8086/query --data-urlencode "q=drop subscription kapacitor on _internal.default"

That produced the following output:

    ERR: error parsing query: found DEFAULT, expected identifier at line 1, char 42

So I found this documentation (https://github.com/influxdata/influxdb/blob/master/influxql/INFLUXQL.md) and added quotes to it:

    curl -G http://localhost:8086/query --data-urlencode 'q=drop subscription kapacitor on "_internal"."default"'

Output:

    ERR: Get http://localhost:8086/query?db=_internal&epoch=ns&q=drop+subscription+kapacitor+on+%22_internal%22.%22default%22: EOF

I didn't care too much about the output, thinking I still had my drop syntax wrong, but when I ran `select * from _internal.monitor` I got this:

    ERR: Get http://localhost:8086/query?db=_internal&epoch=ns&q=select+%2A+from+_internal.monitor: dial tcp 127.0.0.1:8086: connection refused

It turns out influxdb died. Here's what the logs said: https://gist.github.com/robinsk/ad7564ddeb469fdb69e7#file-influxdb-log

Mainly:

    [query] 2016/02/05 09:59:22 DROP SUBSCRIPTION kapacitor ON _internal."default"
    panic: runtime error: invalid memory address or nil pointer dereference

When trying to start the influxdb service again, it fails, with almost the same things in the logs:
https://gist.github.com/robinsk/ad7564ddeb469fdb69e7#file-influxdb_starting-log

Is there any way to restore normal operations on this database, or do I have to restore from backup? This is in a vm in vagrant for testing, so it didn't matter _that_ much this time, but that makes me reluctant of running the same in production.

I've hit the same issue with `0.10.0`. Stack trace here:

```
[query] 2016/02/08 21:46:21 DROP SUBSCRIPTION kapacitor ON _internal."default"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x48 pc=0x56d1b0]

goroutine 12 [running]:
github.com/influxdb/influxdb/services/meta.(*Data).DropSubscription(0xc208c33290, 0xc2083ebec0, 0x9, 0xc2083ebe78, 0x7, 0xc2083ebea0, 0x9, 0x0, 0x0)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/services/meta/data.go:606 +0xc0
github.com/influxdb/influxdb/services/meta.(*storeFSM).applyDropSubscriptionCommand(0xc20800c1b0, 0xc2081ee540, 0x0, 0x0)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/services/meta/store_fsm.go:404 +0x1c3
github.com/influxdb/influxdb/services/meta.func·009(0x0, 0x0)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/services/meta/store_fsm.go:66 +0x6f9
github.com/influxdb/influxdb/services/meta.(*storeFSM).Apply(0xc20800c1b0, 0xc2083c8200, 0x0, 0x0)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/services/meta/store_fsm.go:94 +0x2de
github.com/hashicorp/raft.(*Raft).runFSM(0xc208084380)
        /tmp/tmp.r1Nea5VXEw/src/github.com/hashicorp/raft/raft.go:564 +0xdfc
github.com/hashicorp/raft.*Raft.(github.com/hashicorp/raft.runFSM)·fm()
        /tmp/tmp.r1Nea5VXEw/src/github.com/hashicorp/raft/raft.go:253 +0x27
github.com/hashicorp/raft.func·011()
        /tmp/tmp.r1Nea5VXEw/src/github.com/hashicorp/raft/state.go:152 +0x51
created by github.com/hashicorp/raft.(*raftState).goFunc
        /tmp/tmp.r1Nea5VXEw/src/github.com/hashicorp/raft/state.go:153 +0xe3
```

When trying to restart the database, it crashes again (I'm assuming due to corruption):

```
2016/02/08 21:50:15 InfluxDB starting, version 0.10.0, branch 0.10.0, commit b8bb32ecad9808ef00219e7d2469514890a0987a, built 2016-02-04T17:06:04.850564
2016/02/08 21:50:15 Go version go1.4.3, GOMAXPROCS set to 1
2016/02/08 21:50:15 Using configuration at: /etc/influxdb/influxdb.conf
[meta] 2016/02/08 21:50:15 Starting meta service
[meta] 2016/02/08 21:50:15 Listening on HTTP: [::]:8091
[metastore] 2016/02/08 21:50:15 Using data dir: /var/lib/influxdb/meta
[metastore] 2016/02/08 21:50:15 Node at kapacitor1:8088 [Follower]
[metastore] 2016/02/08 21:50:17 Node at kapacitor1:8088 [Leader]. peers=[kapacitor1:8088]
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x48 pc=0x56d1b0]

goroutine 12 [running]:
github.com/influxdb/influxdb/services/meta.(*Data).DropSubscription(0xc20800d8c0, 0xc2080f0fc0, 0x9, 0xc2080f0fa9, 0x7, 0xc2080f0fa0, 0x9, 0x0, 0x0)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/services/meta/data.go:606 +0xc0
github.com/influxdb/influxdb/services/meta.(*storeFSM).applyDropSubscriptionCommand(0xc20800c240, 0xc208126270, 0x0, 0x0)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/services/meta/store_fsm.go:404 +0x1c3
github.com/influxdb/influxdb/services/meta.func·009(0x0, 0x0)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/services/meta/store_fsm.go:66 +0x6f9
github.com/influxdb/influxdb/services/meta.(*storeFSM).Apply(0xc20800c240, 0xc2080e5940, 0x0, 0x0)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/services/meta/store_fsm.go:94 +0x2de
github.com/hashicorp/raft.(*Raft).runFSM(0xc2080c2380)
        /tmp/tmp.r1Nea5VXEw/src/github.com/hashicorp/raft/raft.go:564 +0xdfc
github.com/hashicorp/raft.*Raft.(github.com/hashicorp/raft.runFSM)·fm()
        /tmp/tmp.r1Nea5VXEw/src/github.com/hashicorp/raft/raft.go:253 +0x27
github.com/hashicorp/raft.func·011()
        /tmp/tmp.r1Nea5VXEw/src/github.com/hashicorp/raft/state.go:152 +0x51
created by github.com/hashicorp/raft.(*raftState).goFunc
        /tmp/tmp.r1Nea5VXEw/src/github.com/hashicorp/raft/state.go:153 +0xe3
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,7 @@
 - [#5594](https://github.com/influxdata/influxdb/pull/5594): Fix missing url params on lease redirect - @oldmantaiter
 - [#5376](https://github.com/influxdata/influxdb/pull/5376): Fix golint issues in models package. @nuss-justin
 - [#5535](https://github.com/influxdata/influxdb/pull/5535): Update README for referring to Collectd
+- [#5590](https://github.com/influxdata/influxdb/pull/5590): Fix panic when dropping subscription for unknown retention policy.
 
 ## v0.10.0 [2016-02-04]
 

--- a/services/meta/data.go
+++ b/services/meta/data.go
@@ -574,8 +574,7 @@ func (data *Data) CreateSubscription(database, rp, name, mode string, destinatio
 	rpi, err := data.RetentionPolicy(database, rp)
 	if err != nil {
 		return err
-	}
-	if rpi == nil {
+	} else if rpi == nil {
 		return influxdb.ErrRetentionPolicyNotFound(rp)
 	}
 
@@ -601,6 +600,8 @@ func (data *Data) DropSubscription(database, rp, name string) error {
 	rpi, err := data.RetentionPolicy(database, rp)
 	if err != nil {
 		return err
+	} else if rpi == nil {
+		return influxdb.ErrRetentionPolicyNotFound(rp)
 	}
 
 	for i := range rpi.Subscriptions {

--- a/services/meta/data.go
+++ b/services/meta/data.go
@@ -89,7 +89,7 @@ func (data *Data) setDataNode(nodeID uint64, host, tcpHost string) error {
 	return nil
 }
 
-// DeleteNode removes a node from the metadata.
+// DeleteDataNode removes a node from the metadata.
 func (data *Data) DeleteDataNode(id uint64) error {
 	// Node has to be larger than 0 to be real
 	if id == 0 {

--- a/services/meta/errors.go
+++ b/services/meta/errors.go
@@ -69,9 +69,6 @@ var (
 	// ErrReplicationFactorTooLow is returned when the replication factor is not in an
 	// acceptable range.
 	ErrReplicationFactorTooLow = errors.New("replication factor must be greater than 0")
-
-	// ErrRetentionPolicyNotFound is returned when an expected retention policy can't be found.
-	ErrRetentionPolicyNotFound = errors.New("retention policy not found")
 )
 
 var (

--- a/services/meta/service_test.go
+++ b/services/meta/service_test.go
@@ -13,6 +13,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/influxdb/influxdb"
+
 	"github.com/influxdb/influxdb/influxql"
 	"github.com/influxdb/influxdb/services/meta"
 	"github.com/influxdb/influxdb/tcp"
@@ -566,7 +568,7 @@ func TestMetaService_ContinuousQueries(t *testing.T) {
 	}
 }
 
-func TestMetaService_Subscriptions(t *testing.T) {
+func TestMetaService_Subscriptions_Create(t *testing.T) {
 	t.Parallel()
 
 	d, s, c := newServiceAndClient()
@@ -598,36 +600,139 @@ func TestMetaService_Subscriptions(t *testing.T) {
 	res := c.ExecuteStatement(mustParseStatement(`SHOW SUBSCRIPTIONS`))
 	if res.Err != nil {
 		t.Fatal(res.Err)
+	} else if got, exp := len(res.Series), 1; got != exp {
+		t.Fatalf("unexpected response.\n\ngot: %d series\nexp: %d\n", got, exp)
+	}
+
+	// Create another subscription.
+	if res := c.ExecuteStatement(mustParseStatement(`CREATE SUBSCRIPTION sub1 ON db0."default" DESTINATIONS ALL 'udp://example.com:6060'`)); res.Err != nil {
+		t.Fatal(res.Err)
+	}
+
+	// The subscriptions are correctly created.
+	if res = c.ExecuteStatement(mustParseStatement(`SHOW SUBSCRIPTIONS`)); res.Err != nil {
+		t.Fatal(res.Err)
+	}
+
+	exp := `{"series":[{"name":"db0","columns":["retention_policy","name","mode","destinations"],"values":[["default","sub0","ALL",["udp://example.com:9090"]],["default","sub1","ALL",["udp://example.com:6060"]]]}]}`
+	got := mustMarshalJSON(res)
+	if got != exp {
+		t.Fatalf("unexpected response.\n\ngot: %s\nexp: %s\n", exp, got)
+	}
+
+	// // Re-create a subscription
+	// if res := c.ExecuteStatement(mustParseStatement(`DROP SUBSCRIPTION sub1 ON db0."default"`)); res.Err != nil {
+	// 	t.Fatal(res.Err)
+	// }
+
+	// res = c.ExecuteStatement(mustParseStatement(`SHOW SUBSCRIPTIONS`))
+	// if res.Err != nil {
+	// 	t.Fatal(res.Err)
+	// }
+
+}
+
+func TestMetaService_Subscriptions_Show(t *testing.T) {
+	t.Parallel()
+
+	d, s, c := newServiceAndClient()
+	defer os.RemoveAll(d)
+	defer s.Close()
+	defer c.Close()
+
+	// Create a database to use
+	if res := c.ExecuteStatement(mustParseStatement("CREATE DATABASE db0")); res.Err != nil {
+		t.Fatal(res.Err)
+	}
+	db, err := c.Database("db0")
+	if err != nil {
+		t.Fatal(err)
+	} else if db.Name != "db0" {
+		t.Fatalf("db name wrong: %s", db.Name)
+	}
+
+	// SHOW SUBSCRIPTIONS returns no subscriptions when there are none.
+	res := c.ExecuteStatement(mustParseStatement(`SHOW SUBSCRIPTIONS`))
+	if res.Err != nil {
+		t.Fatal(res.Err)
+	} else if got, exp := len(res.Series), 0; got != exp {
+		t.Fatalf("got %d series, expected %d", got, exp)
+	}
+
+	// Create a subscription.
+	if res = c.ExecuteStatement(mustParseStatement(`CREATE SUBSCRIPTION sub0 ON db0."default" DESTINATIONS ALL 'udp://example.com:9090'`)); res.Err != nil {
+		t.Fatal(res.Err)
+	}
+
+	// SHOW SUBSCRIPTIONS returns the created subscription.
+	if res = c.ExecuteStatement(mustParseStatement(`SHOW SUBSCRIPTIONS`)); res.Err != nil {
+		t.Fatal(res.Err)
+	} else if got, exp := len(res.Series), 1; got != exp {
+		t.Fatalf("got %d series, expected %d", got, exp)
 	}
 
 	exp := `{"series":[{"name":"db0","columns":["retention_policy","name","mode","destinations"],"values":[["default","sub0","ALL",["udp://example.com:9090"]]]}]}`
 	got := mustMarshalJSON(res)
-	if exp != got {
-		t.Fatalf("unexpected response.\n\nexp: %s\ngot: %s\n", exp, got)
+	if got != exp {
+		t.Fatalf("unexpected response.\n\ngot: %s\nexp: %s\n", got, exp)
 	}
+}
 
-	// Create a couple more subscriptions
-	if res := c.ExecuteStatement(mustParseStatement(`CREATE SUBSCRIPTION sub1 ON db0."default" DESTINATIONS ALL 'udp://example.com:6060'`)); res.Err != nil {
+func TestMetaService_Subscriptions_Drop(t *testing.T) {
+	t.Parallel()
+
+	d, s, c := newServiceAndClient()
+	defer os.RemoveAll(d)
+	defer s.Close()
+	defer c.Close()
+
+	// Create a database to use
+	if res := c.ExecuteStatement(mustParseStatement("CREATE DATABASE db0")); res.Err != nil {
 		t.Fatal(res.Err)
 	}
-	if res := c.ExecuteStatement(mustParseStatement(`CREATE SUBSCRIPTION sub2 ON db0."default" DESTINATIONS ALL 'udp://example.com:7070'`)); res.Err != nil {
+	db, err := c.Database("db0")
+	if err != nil {
+		t.Fatal(err)
+	} else if db.Name != "db0" {
+		t.Fatalf("db name wrong: %s", db.Name)
+	}
+
+	// DROP SUBSCRIPTION returns ErrSubscriptionNotFound when the
+	// subscription is unknown.
+	res := c.ExecuteStatement(mustParseStatement(`DROP SUBSCRIPTION foo ON db0."default"`))
+	if got, exp := res.Err, meta.ErrSubscriptionNotFound; got.Error() != exp.Error() {
+		t.Fatalf("got: %s, exp: %s", got.Error(), exp)
+	}
+
+	// Create a subscription.
+	if res = c.ExecuteStatement(mustParseStatement(`CREATE SUBSCRIPTION sub0 ON db0."default" DESTINATIONS ALL 'udp://example.com:9090'`)); res.Err != nil {
 		t.Fatal(res.Err)
 	}
 
-	// Re-create a subscription
-	if res := c.ExecuteStatement(mustParseStatement(`DROP SUBSCRIPTION sub1 ON db0."default"`)); res.Err != nil {
-		t.Fatal(res.Err)
+	// DROP SUBSCRIPTION returns an influxdb.ErrDatabaseNotFound when
+	// the database is unknown.
+	res = c.ExecuteStatement(mustParseStatement(`DROP SUBSCRIPTION sub0 ON foo."default"`))
+	if got, exp := res.Err, influxdb.ErrDatabaseNotFound("foo"); got.Error() != exp.Error() {
+		t.Fatalf("got: %s, exp: %s", got.Error(), exp)
 	}
 
-	res = c.ExecuteStatement(mustParseStatement(`SHOW SUBSCRIPTIONS`))
-	if res.Err != nil {
-		t.Fatal(res.Err)
+	// DROP SUBSCRIPTION returns an influxdb.ErrRetentionPolicyNotFound
+	// when the retention policy is unknown.
+	res = c.ExecuteStatement(mustParseStatement(`DROP SUBSCRIPTION sub0 ON db0."foo_policy"`))
+	if got, exp := res.Err, influxdb.ErrRetentionPolicyNotFound("foo_policy"); got.Error() != exp.Error() {
+		t.Fatalf("got: %s, exp: %s", got.Error(), exp)
 	}
 
-	exp = `{"series":[{"name":"db0","columns":["retention_policy","name","mode","destinations"],"values":[["default","sub0","ALL",["udp://example.com:9090"]],["default","sub2","ALL",["udp://example.com:7070"]]]}]}`
-	got = mustMarshalJSON(res)
-	if exp != got {
-		t.Fatalf("unexpected response.\n\nexp: %s\ngot: %s\n", exp, got)
+	// DROP SUBSCRIPTION drops the subsciption if it can find it.
+	res = c.ExecuteStatement(mustParseStatement(`DROP SUBSCRIPTION sub0 ON db0."default"`))
+	if got := res.Err; got != nil {
+		t.Fatalf("got: %s, exp: %s", got.Error(), nil)
+	}
+
+	if res = c.ExecuteStatement(mustParseStatement(`SHOW SUBSCRIPTIONS`)); res.Err != nil {
+		t.Fatal(res.Err)
+	} else if got, exp := len(res.Series), 0; got != exp {
+		t.Fatalf("got %d series, expected %d", got, exp)
 	}
 }
 

--- a/services/meta/service_test.go
+++ b/services/meta/service_test.go
@@ -619,17 +619,6 @@ func TestMetaService_Subscriptions_Create(t *testing.T) {
 	if got != exp {
 		t.Fatalf("unexpected response.\n\ngot: %s\nexp: %s\n", exp, got)
 	}
-
-	// // Re-create a subscription
-	// if res := c.ExecuteStatement(mustParseStatement(`DROP SUBSCRIPTION sub1 ON db0."default"`)); res.Err != nil {
-	// 	t.Fatal(res.Err)
-	// }
-
-	// res = c.ExecuteStatement(mustParseStatement(`SHOW SUBSCRIPTIONS`))
-	// if res.Err != nil {
-	// 	t.Fatal(res.Err)
-	// }
-
 }
 
 func TestMetaService_Subscriptions_Show(t *testing.T) {
@@ -726,7 +715,7 @@ func TestMetaService_Subscriptions_Drop(t *testing.T) {
 	// DROP SUBSCRIPTION drops the subsciption if it can find it.
 	res = c.ExecuteStatement(mustParseStatement(`DROP SUBSCRIPTION sub0 ON db0."default"`))
 	if got := res.Err; got != nil {
-		t.Fatalf("got: %s, exp: %s", got.Error(), nil)
+		t.Fatalf("got: %s, exp: %v", got.Error(), nil)
 	}
 
 	if res = c.ExecuteStatement(mustParseStatement(`SHOW SUBSCRIPTIONS`)); res.Err != nil {
@@ -883,7 +872,7 @@ func TestMetaService_CommandAgainstNonLeader(t *testing.T) {
 
 	cfgs := make([]*meta.Config, 3)
 	srvs := make([]*testService, 3)
-	for i, _ := range cfgs {
+	for i := range cfgs {
 		c := newConfig()
 
 		cfgs[i] = c
@@ -926,7 +915,7 @@ func TestMetaService_FailureAndRestartCluster(t *testing.T) {
 
 	cfgs := make([]*meta.Config, 3)
 	srvs := make([]*testService, 3)
-	for i, _ := range cfgs {
+	for i := range cfgs {
 		c := newConfig()
 
 		cfgs[i] = c
@@ -1208,18 +1197,18 @@ func TestMetaService_PersistClusterIDAfterRestart(t *testing.T) {
 	}
 	defer c.Close()
 
-	id_after := c.ClusterID()
-	if id_after == 0 {
+	idAfter := c.ClusterID()
+	if idAfter == 0 {
 		t.Fatal("cluster ID can't be zero")
-	} else if id_after != id {
-		t.Fatalf("cluster id not the same: %d, %d", id_after, id)
+	} else if idAfter != id {
+		t.Fatalf("cluster id not the same: %d, %d", idAfter, id)
 	}
 }
 
 func TestMetaService_Ping(t *testing.T) {
 	cfgs := make([]*meta.Config, 3)
 	srvs := make([]*testService, 3)
-	for i, _ := range cfgs {
+	for i := range cfgs {
 		c := newConfig()
 
 		cfgs[i] = c

